core/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java

```diff
@@ -150,24 +155,34 @@ public AllocateUnassignedDecision makeAllocationDecision(final ShardRouting unas
         final RoutingNodes routingNodes = allocation.routingNodes();
         final boolean explain = allocation.debugDecision();
         // pre-check if it can be allocated to any node that currently exists, so we won't list the store for it for nothing
-        Tuple<Decision, Map<String, Decision>> allocateDecision = canBeAllocatedToAtLeastOneNode(unassignedShard, allocation, explain);
-        if (allocateDecision.v1().type() != Decision.Type.YES) {
+        Tuple<Decision, Map<String, NodeAllocationResult>> result = canBeAllocatedToAtLeastOneNode(unassignedShard, allocation);
+        Decision allocateDecision = result.v1();
+        if (allocateDecision.type() != Decision.Type.YES) {
+            // only return early if we are not in explain mode, because if we are in explain mode,
```

Comment:
I don't understand this explanation. we return early here, even in explain mode?